eppo_core/src/ufc/eval.rs

```diff
@@ -153,85 +258,108 @@ impl Flag {
                            subject_key,
                            variation_key:display = split.variation_key;
                            "internal: unable to convert Value to AssignmentValue");
-                Error::ConfigurationError
+                FlagEvaluationError::ConfigurationError
             })?;
 
-        let event = if allocation.do_log {
-            Some(AssignmentEvent {
-                feature_flag: self.key.clone(),
-                allocation: allocation.key.clone(),
-                experiment: format!("{}-{}", self.key, allocation.key),
-                variation: variation.key.clone(),
-                subject: subject_key.to_owned(),
-                subject_attributes: subject_attributes.clone(),
-                timestamp: now.to_rfc3339(),
-                meta_data: [(
-                    "eppoCoreVersion".to_owned(),
-                    env!("CARGO_PKG_VERSION").to_owned(),
-                )]
-                .into(),
-                extra_logging: split.extra_logging.clone(),
-            })
-        } else {
-            None
-        };
-
-        Ok(Some(Assignment {
+        let event = allocation.do_log.then(|| AssignmentEvent {
+            feature_flag: self.key.clone(),
+            allocation: allocation.key.clone(),
+            experiment: format!("{}-{}", self.key, allocation.key),
+            variation: variation.key.clone(),
+            subject: subject_key.to_owned(),
+            subject_attributes: subject_attributes.clone(),
+            timestamp: now.to_rfc3339(),
+            meta_data: [(
+                "eppoCoreVersion".to_owned(),
+                env!("CARGO_PKG_VERSION").to_owned(),
+            )]
+            .into(),
+            extra_logging: split.extra_logging.clone(),
+            evaluation_details: None,
+        });
+
+        Ok(Assignment {
             value: assignment_value,
             event,
-        }))
+        })
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(super) enum AllocationNonMatchReason {
+    BeforeStartDate,
+    AfterEndDate,
+    FailingRules,
+    TrafficExposureMiss,
+}
+
 impl Allocation {
-    fn get_matching_split(
+    fn get_matching_split<V: EvalAllocationVisitor>(
         &self,
+        visitor: &mut V,
         subject_key: &str,
         subject_attributes_with_id: &Attributes,
-        sharder: &impl Sharder,
         total_shards: u64,
         now: Timestamp,
-    ) -> Option<&Split> {
-        if self.is_allowed_by_time(now) && self.is_allowed_by_rules(subject_attributes_with_id) {
-            self.splits
-                .iter()
-                .find(|split| split.matches(subject_key, sharder, total_shards))
-        } else {
-            None
+    ) -> Result<&Split, AllocationNonMatchReason> {
+        if self.start_at.is_some_and(|t| now < t) {
+            return Err(AllocationNonMatchReason::BeforeStartDate);
```

Comment:
Yes. `Error<T, E>` is just a normal rust type for when one branch is returning a value (`Ok(value)`) and another one fails to produce the value for any reason (`Err(err)`). Unless the reason is trivial, in which case one would use `Option<T>` with `Some(value)` and `None`. This is actually the reason why the function signature changed from `Option` to `Result`—because we want to know the reason